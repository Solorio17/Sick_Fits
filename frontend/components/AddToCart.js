import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query, Mutation } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';

const ADD_TO_CART_MUTATION = gql`
    mutation addToCart($id: ID!){
        addToCart(id: $id){
            id
            quantity
        }
    }
`;

class AddToCart extends Component{
    render(){
        const { id } = this.props
        return(
            <Mutation 
                mutation={ADD_TO_CART_MUTATION} 
                variables={{
                    id: id
                }}
                refetchQueries={[{ query: CURRENT_USER_QUERY }]}
            >
                {(addToCart, { error, loading}) => <button className="adding" onClick={addToCart} disabled={loading}>Add{loading && 'ing'} to cart</button>}
            </Mutation>
        )
    }
}

export default AddToCart;
export { ADD_TO_CART_MUTATION };