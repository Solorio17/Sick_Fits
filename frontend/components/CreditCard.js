import React, { Component } from 'react';
import gql from 'graphql-tag';
import StripeCheckout from 'react-stripe-checkout';
import { Mutation } from 'react-apollo';
import Router from 'next/router';
import NProgress from 'nprogress';
import propTypes from 'prop-types';
import calcTotalPrice from '../lib/calcTotalPrice';
import Error from './ErrorMessage';
import User, { CURRENT_USER_QUERY } from './User';

const CREATE_ORDER_MUTATION = gql`
    mutation createOrder($token: String!){
        createOrder(token: $token){
            id
            charge
            total
            items{
                id
                title
            }
        }
    }
`;

function totalItems(cart){
    return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)
}

class CreditCard extends Component{
    onToken = async (res, createOrder) =>{
        NProgress.start();
        // console.log(res.id)
        //manually call mutation once we have stripe token
        const order = await createOrder({
            variables: {
                token: res.id
            },
        }).catch(err => {
            alert(err.message)
        });
        Router.push({
            pathname: "/order",
            query: { id: order.data.createOrder.id }
        })
    }
    render(){
        return(
            <User>
                {({data: { me }, loading }) => {

                if(loading) return null;
                
                return (
                    <Mutation mutation={CREATE_ORDER_MUTATION} refetchQueries={[{ query: CURRENT_USER_QUERY }]}> 
                        {(createOrder) => (
                        
                        <StripeCheckout
                            amount={calcTotalPrice(me.cart)}
                            name="Sick Fits"
                            description={`Total items: ${totalItems(me.cart)}`}
                            image={me.cart.length && me.cart[0].item && me.cart[0].item.image}
                            stripeKey="pk_test_LPv396GsOHWwT37lbslxNgW3"
                            currency="USD"
                            email={me.email}
                            token={res => this.onToken(res, createOrder)}
                        >
                            {this.props.children}
                        </StripeCheckout>
                        )}
                    </Mutation>
                )}}
            </User>
        )
    }
}

export default CreditCard;
export { CREATE_ORDER_MUTATION };