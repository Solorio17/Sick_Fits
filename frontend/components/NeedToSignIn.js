import { Query } from 'react-apollo';
import { CURRENT_USER_QUERY } from './User';
import Signin from './Signin';

const NeedToSignIn = props =>{
    return(
        <Query query={CURRENT_USER_QUERY}>
            {({data, loading}) => {
                if(loading) return <p>Loading...</p>
                if(!data.me){
                    return(
                        <div>
                            <h3>Please sign in first!</h3>
                            <Signin/>
                        </div>
                    )
                }
                return props.children
            }}
        </Query>
    )
}

export default NeedToSignIn;