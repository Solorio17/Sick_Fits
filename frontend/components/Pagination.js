import React from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import { perPage } from '../config';
import Head from 'next/head'
import Link from 'next/link';
import PaginationStyles from '../components/styles/PaginationStyles';


const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        itemsConnection{
            aggregate{
                count
            }
        }
    }
`;


const Pagination = props =>{
    return(
            <Query query={PAGINATION_QUERY}>
                {({ data, loading, error}) =>{
                    if(loading) return <p>Loading...</p>
                    const count = data.itemsConnection.aggregate.count;
                    const pages = Math.ceil(count / perPage);
                    const page = props.page;
                    const previous = "<< Prev";
                    const next = "Next >>";

                    return(
                        <PaginationStyles>
                            <Head>
                                <title>Sick Fits - Page {props.page} of {pages}</title>
                            </Head>
                            <Link 
                                prefetch
                                href={{
                                pathname: 'items',
                                query: {page: page - 1}
                            }}>
                            <a className="prev" aria-disabled={page <= 1}>{previous}</a>
                            </Link>
                            <p>Page {page} of {pages}</p>
                            <p>{count} Items total</p>
                            <Link 
                                prefetch
                                href={{
                                pathname: 'items',
                                query: {page: page + 1}
                            }}>
                            <a className="next" aria-disabled={page >= pages}>{next}</a>
                            </Link>
                        </PaginationStyles>
                    )
                }}
            </Query>
    )
}

export default Pagination;