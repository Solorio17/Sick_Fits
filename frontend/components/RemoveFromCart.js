import React, { Component } from 'react';
import styled from 'styled-components';
import propTypes from 'prop-types';
import gql from 'graphql-tag';
import { CURRENT_USER_QUERY } from './User';
import { Mutation } from 'react-apollo';

const REMOVE_FROM_CART_MUTATION = gql`
    mutation removeFromCart($id: ID!){
        removeFromCart(id: $id){
            id
        }
    }
`;

const DeleteItemButton = styled.button`
    font-size: 18px;
    background: none;
    border: 0;
    &:hover{
        color: ${props => props.theme.red};
        cursor: pointer
    }
`;

class RemoveFromCart extends Component{
    static propTypes = { id: propTypes.string.isRequired };

    update = ( cache, payload ) => {
        //read cache
        const data = cache.readQuery({
            query: CURRENT_USER_QUERY
        })
        //remove the item from cart
        const cartItemId = payload.data.removeFromCart.id;
        data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
        //write it back to the cache
        cache.writeQuery({ query: CURRENT_USER_QUERY, data: data})
    }

    render(){
        return(
            <Mutation 
                mutation={REMOVE_FROM_CART_MUTATION} 
                variables={{id: this.props.id}} 
                update={this.update}
                optimisticResponse={{
                    __typename: 'Mutation',
                    removeFromCart: {
                        __typename: 'CartItem',
                        id: this.props.id
                    }
                }}
            >
                {(removeFromCart, { loading, error }) => (
                    <DeleteItemButton 
                        title="deleteItem" 
                        onClick={() => {removeFromCart().catch(error => alert(error.message))}} 
                        disabled={loading}
                    > &times; </DeleteItemButton>
                )}
            </Mutation>
        )
    }
}

export default RemoveFromCart;