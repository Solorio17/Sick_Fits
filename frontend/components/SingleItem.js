import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Error from '../components/ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';

const SingleItemStyle = styled.div`
    max-width: 1200px;
    margin: 2rem auto;
    box-shadow: ${props => props.theme.bs};
    display: grid;
    grid-auto-columns: 1fr;
    grid-auto-flow: column;
    min-height: 800px;
    img{
        width: 100%;
        height: 100%;
        object-fit: contain;
    }
    .details{
        text-align: center;
    }
    .details h2{
        font-style: italic;
    }
    .details p span{
        font-weight: bold;
    }
`;

const SINGLE_ITEM_QUERY = gql`
    query SINGLE_ITEM_QUERY($id: ID!){
        item(where: { id: $id }){
            id
            title
            description
            price
            largeImage
        }
    }
`;

class SingleItem extends Component{
    render(){
        return(
            <Query query={SINGLE_ITEM_QUERY} variables={{
                id: this.props.id
            }}>
            {({error, loading, data}) => {
                if(error) return <Error error={error}/>
                if(loading) return <p>Loading...</p>
                if(!data.item) return <p>No item found for {this.props.id}</p>
                const item = data.item;
                return(
                    <SingleItemStyle>
                        <Head>
                            <title>Sick Fits | {item.title}</title>
                        </Head>
                        <img src={item.largeImage} alt={item.title}/>
                        <div className="details">
                            <h2>{item.title}</h2>
                            <p><span>More about this item: </span>{item.description}</p>
                            <h3>{"$" + item.price /100}</h3>
                        </div>
                    </SingleItemStyle>
                )
            }}
            </Query>
        )
    }
}

export default SingleItem;
export { SINGLE_ITEM_QUERY };