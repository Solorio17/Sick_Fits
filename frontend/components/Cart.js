import React from 'react';
import gql from 'graphql-tag';
import { Query, Mutation } from 'react-apollo';
import calcTotalPrice from '../lib/calcTotalPrice';
import formatMoney from '../lib/formatMoney';
import { adopt } from 'react-adopt';
import CartStyles from './styles/CartStyles';
import SickButton from './styles/SickButton';
import Supreme from './styles/Supreme';
import CloseButton from './styles/CloseButton';
import CartItem from './CartItem';
import User from './User';
import CreditCard from './CreditCard';

const LOCAL_STATE_QUERY = gql`
    query{
        cartOpen @client
    }
`;

const TOGGLE_CART_MUTATION = gql`
    mutation{
        toggleCart @client
    }
`;

const Composed = adopt({
    user: ({ render }) => <User>{render}</User>,
    toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
    localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>
})

const Cart = () => {
    return(
        <Composed>
            {({user, toggleCart, localState}) => {
                const me = user.data.me;
                if(!me) return null
                return(
                    <CartStyles open={localState.data.cartOpen}>
                        <header>
                            <CloseButton onClick={toggleCart} title="close">X</CloseButton>
                            <Supreme>{me.name}'s Cart</Supreme>
                            <p>You have {me.cart.length} different item{me.cart.length === 1 ? '' : 's'} in your cart</p>
                        </header>
                        
                        <ul>{me.cart.map(cartItem => <CartItem key={cartItem.id} cartItem={cartItem}/>)}</ul>

                        <footer>
                            <p>Total: {formatMoney(calcTotalPrice(me.cart))}</p>
                            <CreditCard>
                                <SickButton>Checkout</SickButton>
                            </CreditCard>
                        </footer>
                    </CartStyles>
                )
            }}
        </Composed>
    )
}

export default Cart;
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION };