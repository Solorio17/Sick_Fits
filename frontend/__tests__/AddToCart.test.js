import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import AddToCart, {ADD_TO_CART_MUTATION} from '../components/AddToCart';
import { fakeCartItem, fakeUser } from '../lib/testUtils';
import { ApolloConsumer} from 'react-apollo';
import {CURRENT_USER_QUERY} from '../components/User';

const mocks = [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: []
                }
            }
        }
    },
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: {
                    ...fakeUser(),
                    cart: [fakeCartItem()]
                }
            }
        }
    },
    {
        request: { query: ADD_TO_CART_MUTATION, variables: {id: 'abc123'} },
        result: {
            data: {
                addToCart: {
                    ...fakeCartItem(),
                    quantity: 1
                }
            }
        }
    }
]

describe('AddToCart component', () => {
  it('renders and matches the snapshot', async()=>{
    const wrapper = mount(
        <MockedProvider mocks={mocks}>
            <AddToCart id={'645vdu6bf7iy'}/>
        </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
    
  });

  it('adds an item to the cart when clicked', async()=>{
    let apolloClient;
    const wrapper = mount(
        <MockedProvider mocks={mocks}>
            <ApolloConsumer>
                {(client)=>{
                    apolloClient = client;
                    return <AddToCart id={'abc123'}/>
                }}
            </ApolloConsumer>
        </MockedProvider>
    );
    await wait();
    wrapper.update();
    const { data: { me } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    // console.log(me)
    expect(me.cart).toHaveLength(0);
  
    //add item to cart
    wrapper.find('button.adding').simulate('click');
    await wait();
    wrapper.update()
    //check if item is in the cart
    const { data: { me: me2 } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    await wait();
    wrapper.update()

    expect(me2.cart).toHaveLength(1);
    // expect(me2.cart[0].id).toBe('2q3rwgr');
    // expect(me2.cart[0].quantity).toBe(3);
  });

  it('changes from add to adding when clicked', async()=>{
    const wrapper = mount(
        <MockedProvider mocks={mocks}>
            <AddToCart id="abc123" />
        </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.text()).toContain('Add to cart');
    wrapper.find('button').simulate('click');
    expect(wrapper.text()).toContain('Adding to cart');
  });
});
