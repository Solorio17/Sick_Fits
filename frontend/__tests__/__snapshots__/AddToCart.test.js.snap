// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCart component renders and matches the snapshot 1`] = `
<button
  className="adding"
  disabled={false}
  onClick={[Function]}
>
  Add
   to cart
</button>
`;
